File: haskell-server/src/EagerWords/Common/Domain/PlayerType.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}

module EagerWords.Common.Domain.PlayerType (
    PlayerType (..)
  , userIndex, machineIndex
  , playerTypeIndex
) where

import GHC.Generics
import Data.Aeson (FromJSON, ToJSON)
import Control.DeepSeq (NFData)
import qualified Data.Aeson as Aeson
import qualified Data.ByteString.Lazy.Char8 as BC

-- | Types of players.
data PlayerType = User | Machine
  deriving (Show, Read, Eq, NFData, Generic)

instance FromJSON PlayerType
instance ToJSON PlayerType

-- | Index of user player - for indexing into an array of user/machine.
userIndex = 0 :: Int
-- | Index of machine player - for indexing into an array of user/machine.
machineIndex = 1 :: Int

-- | Get the index of a player for indexing into an array of user/machine.
playerTypeIndex :: PlayerType -> Int
playerTypeIndex User = userIndex
playerTypeIndex Machine = machineIndex

File: scala-server/app/com/bolour/eagerwords/common/domain/PlayerType.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.common.domain

object PlayerType {

  val UserString = "User"
  val MachineString = "Machine"

  val RandomString = "Random"

  sealed abstract class PlayerType

  object UserPlayer extends PlayerType {
    override def toString = UserString
  }

  object MachinePlayer extends PlayerType {
    override def toString = MachineString
  }

  object RandomPlayer extends PlayerType {
    override def toString = RandomString
  }

  def playerIndex(playerType: PlayerType): Int = playerType match {
    case UserPlayer => 0
    case MachinePlayer => 1
    case RandomPlayer => throw new IllegalArgumentException(s"cannot index random player")
  }

  def nextPlayerType(playerType: PlayerType): PlayerType =
    playerType match {
      case UserPlayer => MachinePlayer
      case MachinePlayer => UserPlayer
      case RandomPlayer => throw new IllegalArgumentException(s"next player is meaningless for random player")
    }

  def fromString(asString: String): PlayerType = {
    asString match {
      case UserString => PlayerType.UserPlayer
      case MachineString => PlayerType.MachinePlayer
      case RandomString => PlayerType.RandomPlayer
      case _ => throw new RuntimeException(s"unrecognized player type: ${asString}")
    }
  }
}

